File: Java/Complex Flow Examples/CreateInvoiceFromStructuredData.java
/*
 * Generate a tagged invoice from JSON and CSV input using pdfRest.
 *
 * Set PDFREST_API_KEY in .env or the environment before running this sample.
 * It reads metadata.json, style.json, and line-items.csv from invoice-data,
 * then downloads the PDF beside this class. Run from this directory with:
 * mvn package && java -cp target/pdf-rest-api-samples-1.0-SNAPSHOT-jar-with-dependencies.jar CreateInvoiceFromStructuredData
 */

import io.github.cdimascio.dotenv.Dotenv;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import org.json.JSONArray;
import org.json.JSONObject;

public class CreateInvoiceFromStructuredData {
  private static final Path DATA_DIR = Path.of("invoice-data");
  private static final Path OUTPUT = Path.of("invoice-from-structured-data.pdf");
  private static final String DEFAULT_URL = "https://api.pdfrest.com";
  private static final String DEFAULT_KEY = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx";

  private static final OkHttpClient CLIENT = new OkHttpClient.Builder()
      .readTimeout(120, TimeUnit.SECONDS).build();

  private static JSONObject postJson(String baseUrl, String key, String endpoint, JSONObject payload) throws IOException {
    RequestBody body = RequestBody.create(payload.toString(), MediaType.parse("application/json"));
    Request request = new Request.Builder().url(baseUrl + "/" + endpoint).header("Api-Key", key)
        .header("Accept", "application/json").post(body).build();
    try (Response response = CLIENT.newCall(request).execute()) { return result(response, endpoint); }
  }

  private static JSONObject postMultipart(String baseUrl, String key, String endpoint, List<Part> parts) throws IOException {
    MultipartBody.Builder builder = new MultipartBody.Builder().setType(MultipartBody.FORM);
    for (Part part : parts) {
      if (part.file != null) builder.addFormDataPart(part.name, part.file.getFileName().toString(), RequestBody.create(part.file.toFile(), MediaType.parse("application/octet-stream")));
      else builder.addFormDataPart(part.name, part.value);
    }
    Request request = new Request.Builder().url(baseUrl + "/" + endpoint).header("Api-Key", key)
        .header("Accept", "application/json").post(builder.build()).build();
    try (Response response = CLIENT.newCall(request).execute()) { return result(response, endpoint); }
  }

  private static JSONObject result(Response response, String endpoint) throws IOException {
    String text = response.body() == null ? "" : response.body().string();
    System.out.println(endpoint + ": " + response.code());
    if (!response.isSuccessful()) throw new IOException(endpoint + " failed: " + text);
    return new JSONObject(text);
  }

  private static String rgb(JSONArray values) { List<String> result = new ArrayList<>(); for (Object value : values) result.add(value.toString()); return String.join(",", result); }
  private static String money(double value) { return String.format(java.util.Locale.US, "$%,.2f", value); }

  private static List<String[]> readCsv() throws IOException {
    List<String[]> rows = new ArrayList<>();
    for (String line : Files.readAllLines(DATA_DIR.resolve("line-items.csv"))) {
      if (line.startsWith("itemCode,")) continue;
      List<String> fields = new ArrayList<>(); StringBuilder field = new StringBuilder(); boolean quoted = false;
      for (char character : line.toCharArray()) { if (character == '"') quoted = !quoted; else if (character == ',' && !quoted) { fields.add(field.toString()); field.setLength(0); } else field.append(character); }
      fields.add(field.toString()); if (fields.size() == 4) rows.add(fields.toArray(new String[0]));
    }
    return rows;
  }

  private static JSONArray textObjects(JSONObject metadata, JSONObject style) {
    JSONObject seller = metadata.getJSONObject("seller"), customer = metadata.getJSONObject("customer");
    JSONArray objects = new JSONArray(); String primary = rgb(style.getJSONArray("primaryColorRgb")), muted = rgb(style.getJSONArray("mutedTextColorRgb"));
    addText(objects, style, 1, 405, 724, "INVOICE", 22, primary, 153, "H1", true);
    addText(objects, style, 1, 405, 696, "Invoice " + metadata.getString("invoiceNumber"), 9, muted, 153, "P", true);
    addText(objects, style, 1, 405, 682, "Issued " + metadata.getString("issueDate"), 9, muted, 153, "P", false);
    addText(objects, style, 1, 405, 668, "Due " + metadata.getString("dueDate"), 9, muted, 153, "P", false);
    addText(objects, style, 1, 66, 622, "FROM", 8, primary, 216, "H2", true); addText(objects, style, 1, 66, 606, seller.getString("name"), 9, "34,34,34", 216, "P", true); addText(objects, style, 1, 66, 592, seller.getString("taxId"), 8, muted, 216, "P", false); addText(objects, style, 1, 66, 578, seller.getString("addressLine1"), 8, muted, 216, "P", false); addText(objects, style, 1, 66, 566, seller.getString("city") + ", " + seller.getString("region") + " " + seller.getString("postalCode"), 8, muted, 216, "P", false);
    addText(objects, style, 1, 330, 622, "BILL TO", 8, primary, 216, "H2", true); addText(objects, style, 1, 330, 606, customer.getString("name"), 9, "34,34,34", 216, "P", true); addText(objects, style, 1, 330, 592, customer.getString("addressLine1"), 8, muted, 216, "P", false); addText(objects, style, 1, 330, 578, customer.getString("city") + ", " + customer.getString("region") + " " + customer.getString("postalCode"), 8, muted, 216, "P", false);
    return objects;
  }

  private static void addText(JSONArray objects, JSONObject style, int page, int x, int y, String text, double size, String color, int width, String structure, boolean bold) {
    objects.put(new JSONObject().put("font", bold ? style.getString("boldFont") : style.getString("bodyFont")).put("max_width", width).put("opacity", "1").put("page", String.valueOf(page)).put("rotation", "0").put("text", text).put("text_color_rgb", color).put("text_size", size).put("x", x).put("y", y).put("tag_structure_type", structure));
  }

  private static JSONArray tableObjects(JSONObject metadata, JSONObject style, List<String[]> items) {
    double subtotal = 0; for (String[] item : items) subtotal += Double.parseDouble(item[2]) * Double.parseDouble(item[3]); double tax = Math.round(subtotal * metadata.getDouble("taxRate") * 100) / 100.0; double total = subtotal + tax;
    JSONArray header = new JSONArray(); String[] names = {"Description", "Qty", "Unit Price", "Amount"}; for (int i = 0; i < names.length; i++) { JSONObject cell = new JSONObject().put("text", names[i]).put("tag_structure_type", "TH").put("style", new JSONObject().put("background_color_rgb", style.getJSONArray("primaryColorRgb")).put("text_color_rgb", new JSONArray(List.of(255, 255, 255)))); if (i > 0) cell.getJSONObject("style").put("text_align", "right"); header.put(cell); }
    JSONArray rows = new JSONArray();
    for (String[] item : items) {
      JSONArray cells = new JSONArray();
      cells.put(new JSONObject().put("text", item[1]));
      cells.put(new JSONObject().put("text", item[2]).put("style", new JSONObject().put("text_align", "right")));
      cells.put(new JSONObject().put("text", money(Double.parseDouble(item[3]))).put("style", new JSONObject().put("text_align", "right")));
      cells.put(new JSONObject().put("text", money(Double.parseDouble(item[2]) * Double.parseDouble(item[3]))).put("style", new JSONObject().put("text_align", "right")));
      rows.put(new JSONObject().put("cells", cells));
    }
    JSONArray footer = new JSONArray(); footer.put(summary("Subtotal", subtotal, null)); footer.put(summary("Tax (" + String.format(java.util.Locale.US, "%.2f", metadata.getDouble("taxRate") * 100) + "%)", tax, null)); footer.put(summary("Total", total, style.getJSONArray("accentColorRgb")));
    JSONObject table = new JSONObject().put("page", 1).put("x", 54).put("y", 510).put("width", 504).put("columns", new JSONArray(List.of(new JSONObject().put("width", 276), new JSONObject().put("width", 54), new JSONObject().put("width", 84), new JSONObject().put("width", 90)))).put("continuation_page_top_margin", 85).put("page_bottom_margin", 96).put("final_page_bottom_margin", 164).put("overflow_behavior", "split-row").put("row_split_behavior", "prefer-next-page").put("repeat_header_on_overflow", true).put("show_footer_on_last_page", true).put("tag_structure_type", "Table").put("style", new JSONObject().put("padding", new JSONObject().put("top", 5).put("right", 6).put("bottom", 5).put("left", 6)).put("text_size", style.getDouble("tableHeaderFontSize")).put("text_color_rgb", style.getJSONArray("textColorRgb"))).put("header_rows", new JSONArray(List.of(new JSONObject().put("cells", header)))).put("rows", rows).put("footer_rows", footer);
    return new JSONArray().put(table);
  }

  private static JSONObject summary(String label, double value, JSONArray accent) { JSONObject extra = new JSONObject().put("text_align", "right"); if (accent != null) extra.put("background_color_rgb", accent).put("text_size", 10); return new JSONObject().put("cells", new JSONArray(List.of(new JSONObject().put("text", label).put("col_span", 3).put("style", extra), new JSONObject().put("text", money(value)).put("style", extra)))); }
  private static JSONArray shapes(JSONObject style, int page, boolean footer) { String border = rgb(style.getJSONArray("borderColorRgb")); if (footer) return new JSONArray().put(new JSONObject().put("type", "rectangle").put("page", page).put("x", 54).put("y", 70).put("width", 504).put("height", 104).put("fill_color_rgb", "248,250,251").put("stroke_color_rgb", border).put("stroke_width", 0.5).put("tag_is_artifact", true)); return new JSONArray().put(new JSONObject().put("type", "rectangle").put("page", 1).put("x", 54).put("y", 540).put("width", 240).put("height", 96).put("fill_color_rgb", rgb(style.getJSONArray("accentColorRgb"))).put("stroke_color_rgb", border).put("stroke_width", 0.5).put("tag_is_artifact", true)).put(new JSONObject().put("type", "rectangle").put("page", 1).put("x", 318).put("y", 540).put("width", 240).put("height", 96).put("fill_color_rgb", rgb(style.getJSONArray("accentColorRgb"))).put("stroke_color_rgb", border).put("stroke_width", 0.5).put("tag_is_artifact", true)); }

  public static void main(String[] args) throws Exception {
    Dotenv dotenv = Dotenv.configure().ignoreIfMissing().load(); String apiUrl = (dotenv.get("PDFREST_URL", DEFAULT_URL)).replaceAll("/+$", ""); String apiKey = dotenv.get("PDFREST_API_KEY", DEFAULT_KEY);
    JSONObject metadata = new JSONObject(Files.readString(DATA_DIR.resolve("metadata.json"))), style = new JSONObject(Files.readString(DATA_DIR.resolve("style.json"))); List<String[]> items = readCsv();
    String id = postJson(apiUrl, apiKey, "blank-pdf", new JSONObject().put("page_size", "letter").put("page_count", 1).put("page_orientation", "portrait")).getString("outputId");
    id = postMultipart(apiUrl, apiKey, "pdf-with-added-shapes", List.of(Part.value("id", id), Part.value("shape_objects", shapes(style, 1, false).toString()), Part.value("tag_enabled", "true"))).getString("outputId");
    id = postMultipart(apiUrl, apiKey, "pdf-with-added-text", List.of(Part.value("id", id), Part.value("text_objects", textObjects(metadata, style).toString()), Part.value("tag_enabled", "true"), Part.value("tag_language", "en-US"))).getString("outputId");
    id = postMultipart(apiUrl, apiKey, "pdf-with-added-tables", List.of(Part.value("id", id), Part.value("table_objects", tableObjects(metadata, style, items).toString()), Part.value("tag_enabled", "true"), Part.value("tag_language", "en-US"))).getString("outputId");
    id = postMultipart(apiUrl, apiKey, "pdf-with-added-image", List.of(Part.value("id", id), Part.value("image_objects", new JSONObject().put("image_index", 0).put("page", 1).put("x", 54).put("y", 716).put("width", 200).put("tag_alt_text", "Northstar Sample Supply logo").put("tag_structure_type", "Figure").toString()), Part.file("image_files", DATA_DIR.resolve("northstar-logo.png")), Part.value("tag_enabled", "true"), Part.value("tag_language", "en-US"))).getString("outputId");
    int pageCount = postMultipart(apiUrl, apiKey, "pdf-info", List.of(Part.value("id", id), Part.value("queries", "page_count"))).getInt("page_count");
    id = postMultipart(apiUrl, apiKey, "pdf-with-added-shapes", List.of(Part.value("id", id), Part.value("shape_objects", shapes(style, pageCount, true).toString()), Part.value("tag_enabled", "true"))).getString("outputId");
    JSONArray footer = new JSONArray(); addText(footer, style, pageCount, 66, 156, "Payment terms", 8, rgb(style.getJSONArray("mutedTextColorRgb")), 480, "H2", true); addText(footer, style, pageCount, 66, 142, metadata.getString("paymentTerms"), 7.5, rgb(style.getJSONArray("mutedTextColorRgb")), 480, "P", false); addText(footer, style, pageCount, 66, 112, "Notes", 8, rgb(style.getJSONArray("mutedTextColorRgb")), 480, "H2", true); addText(footer, style, pageCount, 66, 98, metadata.getString("notes"), 7.5, rgb(style.getJSONArray("mutedTextColorRgb")), 480, "P", false);
    for (int page = 1; page <= pageCount; page++) { addText(footer, style, page, 54, 54, "Generated from structured JSON and CSV input with pdfRest.", 7.5, rgb(style.getJSONArray("mutedTextColorRgb")), 400, "P", false); addText(footer, style, page, 490, 54, "Page " + page + " of " + pageCount, 7.5, rgb(style.getJSONArray("mutedTextColorRgb")), 68, "P", false); }
    JSONObject finalResult = postMultipart(apiUrl, apiKey, "pdf-with-added-text", List.of(Part.value("id", id), Part.value("text_objects", footer.toString()), Part.value("tag_enabled", "true"), Part.value("tag_language", "en-US"), Part.value("output", "invoice_from_structured_data")));
    Request download = new Request.Builder().url(apiUrl + "/resource/" + finalResult.getString("outputId") + "?format=file").header("Api-Key", apiKey).get().build(); try (Response response = CLIENT.newCall(download).execute()) { if (!response.isSuccessful()) throw new IOException("Download failed: " + response.code()); Files.write(OUTPUT, response.body().bytes()); }
    System.out.println("Created " + OUTPUT.toAbsolutePath());
  }

  private record Part(String name, String value, Path file) { static Part value(String name, String value) { return new Part(name, value, null); } static Part file(String name, Path file) { return new Part(name, null, file); } }
}

File: PHP/Complex Flow Examples/create-invoice-from-structured-data.php
<?php

/*
 * Generate a tagged invoice from JSON and CSV input using pdfRest.
 *
 * Set PDFREST_API_KEY before running this sample. It reads metadata.json,
 * style.json, and line-items.csv from invoice-data, then downloads the PDF
 * beside this script. Install Guzzle with `composer require guzzlehttp/guzzle`
 * and run
 * with `php create-invoice-from-structured-data.php`.
 */

require 'vendor/autoload.php';

use GuzzleHttp\Client;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Utils;

$apiUrl = rtrim(getenv('PDFREST_URL') ?: 'https://api.pdfrest.com', '/');
$apiKey = getenv('PDFREST_API_KEY') ?: 'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx';
$dataDir = __DIR__ . '/invoice-data';
$outputPath = __DIR__ . '/invoice-from-structured-data.pdf';
$client = new Client(['http_errors' => true]);

function rgb(array $values): string { return implode(',', $values); }
function money(float $value): string { return '$' . number_format($value, 2); }

function postJson(Client $client, string $url, string $apiKey, array $payload): array
{
    $response = $client->send(new Request('POST', $url, ['Accept' => 'application/json', 'Api-Key' => $apiKey, 'Content-Type' => 'application/json'], json_encode($payload)));
    return json_decode($response->getBody(), true, 512, JSON_THROW_ON_ERROR);
}

function postMultipart(Client $client, string $url, string $apiKey, array $fields): array
{
    $multipart = [];
    foreach ($fields as $name => $value) {
        if (is_array($value) && isset($value['path'])) {
            $multipart[] = ['name' => $name, 'contents' => Utils::tryFopen($value['path'], 'r'), 'filename' => $value['name']];
        } else {
            $multipart[] = ['name' => $name, 'contents' => is_string($value) ? $value : json_encode($value)];
        }
    }
    $response = $client->send(new Request('POST', $url, ['Accept' => 'application/json', 'Api-Key' => $apiKey], null), ['multipart' => $multipart]);
    return json_decode($response->getBody(), true, 512, JSON_THROW_ON_ERROR);
}

function textObjects(array $metadata, array $style): array
{
    $seller = $metadata['seller']; $customer = $metadata['customer']; $objects = [];
    $add = function (int $x, int $y, string $text, float $size, string $color, int $width, string $structure = 'P', bool $bold = false) use (&$objects, $style): void {
        $objects[] = ['font' => $bold ? $style['boldFont'] : $style['bodyFont'], 'max_width' => $width, 'opacity' => '1', 'page' => '1', 'rotation' => '0', 'text' => $text, 'text_color_rgb' => $color, 'text_size' => $size, 'x' => $x, 'y' => $y, 'tag_structure_type' => $structure];
    };
    $primary = rgb($style['primaryColorRgb']); $muted = rgb($style['mutedTextColorRgb']);
    $add(405, 724, 'INVOICE', 22, $primary, 153, 'H1', true); $add(405, 696, 'Invoice ' . $metadata['invoiceNumber'], 9, $muted, 153, 'P', true); $add(405, 682, 'Issued ' . $metadata['issueDate'], 9, $muted, 153); $add(405, 668, 'Due ' . $metadata['dueDate'], 9, $muted, 153);
    $add(66, 622, 'FROM', 8, $primary, 216, 'H2', true); $add(66, 606, $seller['name'], 9, '34,34,34', 216, 'P', true); $add(66, 592, $seller['taxId'], 8, $muted, 216); $add(66, 578, $seller['addressLine1'], 8, $muted, 216); $add(66, 566, $seller['city'] . ', ' . $seller['region'] . ' ' . $seller['postalCode'], 8, $muted, 216);
    $add(330, 622, 'BILL TO', 8, $primary, 216, 'H2', true); $add(330, 606, $customer['name'], 9, '34,34,34', 216, 'P', true); $add(330, 592, $customer['addressLine1'], 8, $muted, 216); $add(330, 578, $customer['city'] . ', ' . $customer['region'] . ' ' . $customer['postalCode'], 8, $muted, 216);
    return $objects;
}

function tableObjects(array $metadata, array $style, array $items): array
{
    $subtotal = array_reduce($items, fn(float $sum, array $item): float => $sum + ((float)$item['quantity'] * (float)$item['unitPrice']), 0.0);
    $tax = round($subtotal * $metadata['taxRate'], 2); $total = $subtotal + $tax;
    $headerStyle = ['background_color_rgb' => $style['primaryColorRgb'], 'text_color_rgb' => [255, 255, 255]];
    $headers = ['Description', 'Qty', 'Unit Price', 'Amount']; $header = [];
    foreach ($headers as $index => $text) $header[] = ['text' => $text, 'tag_structure_type' => 'TH', 'style' => $index ? $headerStyle + ['text_align' => 'right'] : $headerStyle];
    $rows = array_map(function (array $item): array { $quantity = (float)$item['quantity']; $unitPrice = (float)$item['unitPrice']; return ['cells' => [['text' => $item['description']], ['text' => $item['quantity'], 'style' => ['text_align' => 'right']], ['text' => money($unitPrice), 'style' => ['text_align' => 'right']], ['text' => money($quantity * $unitPrice), 'style' => ['text_align' => 'right']]]]; }, $items);
    $summary = function (string $label, float $value, array $styleOverride = []): array { return ['cells' => [['text' => $label, 'col_span' => 3, 'style' => ['text_align' => 'right'] + $styleOverride], ['text' => money($value), 'style' => ['text_align' => 'right'] + $styleOverride]]]; };
    return [[
        'page' => 1, 'x' => 54, 'y' => 510, 'width' => 504, 'columns' => [['width' => 276], ['width' => 54], ['width' => 84], ['width' => 90]],
        'continuation_page_top_margin' => 85, 'page_bottom_margin' => 96, 'final_page_bottom_margin' => 164, 'overflow_behavior' => 'split-row', 'row_split_behavior' => 'prefer-next-page', 'repeat_header_on_overflow' => true, 'show_footer_on_last_page' => true, 'tag_structure_type' => 'Table',
        'style' => ['border' => ['top' => ['color_rgb' => $style['borderColorRgb'], 'width' => 0.5], 'right' => ['color_rgb' => $style['borderColorRgb'], 'width' => 0.5], 'bottom' => ['color_rgb' => $style['borderColorRgb'], 'width' => 0.5], 'left' => ['color_rgb' => $style['borderColorRgb'], 'width' => 0.5]], 'padding' => ['top' => 5, 'right' => 6, 'bottom' => 5, 'left' => 6], 'text_size' => $style['tableHeaderFontSize'], 'text_color_rgb' => $style['textColorRgb']],
        'header_rows' => [['cells' => $header]], 'rows' => $rows, 'footer_rows' => [$summary('Subtotal', $subtotal), $summary('Tax (' . number_format($metadata['taxRate'] * 100, 2) . '%)', $tax), $summary('Total', $total, ['background_color_rgb' => $style['accentColorRgb'], 'text_size' => 10])],
    ]];
}

function shapes(array $style, int $page, bool $footer = false): array
{
    $border = ['stroke_color_rgb' => rgb($style['borderColorRgb']), 'stroke_width' => 0.5, 'tag_is_artifact' => true];
    if ($footer) return [['type' => 'rectangle', 'page' => $page, 'x' => 54, 'y' => 70, 'width' => 504, 'height' => 104, 'fill_color_rgb' => '248,250,251'] + $border];
    return [['type' => 'rectangle', 'page' => 1, 'x' => 54, 'y' => 540, 'width' => 240, 'height' => 96, 'fill_color_rgb' => rgb($style['accentColorRgb'])] + $border, ['type' => 'rectangle', 'page' => 1, 'x' => 318, 'y' => 540, 'width' => 240, 'height' => 96, 'fill_color_rgb' => rgb($style['accentColorRgb'])] + $border];
}

$metadata = json_decode(file_get_contents($dataDir . '/metadata.json'), true, 512, JSON_THROW_ON_ERROR);
$style = json_decode(file_get_contents($dataDir . '/style.json'), true, 512, JSON_THROW_ON_ERROR);
$handle = fopen($dataDir . '/line-items.csv', 'r'); $headers = fgetcsv($handle); $items = [];
while (($row = fgetcsv($handle)) !== false) if (count($row) >= count($headers)) $items[] = array_combine($headers, $row);
fclose($handle);

$blank = postJson($client, $apiUrl . '/blank-pdf', $apiKey, ['page_size' => 'letter', 'page_count' => 1, 'page_orientation' => 'portrait']);
$currentId = $blank['outputId'];
$currentId = postMultipart($client, $apiUrl . '/pdf-with-added-shapes', $apiKey, ['id' => $currentId, 'shape_objects' => json_encode(shapes($style, 1)), 'tag_enabled' => 'true'])['outputId'];
$currentId = postMultipart($client, $apiUrl . '/pdf-with-added-text', $apiKey, ['id' => $currentId, 'text_objects' => json_encode(textObjects($metadata, $style)), 'tag_enabled' => 'true', 'tag_language' => 'en-US'])['outputId'];
$currentId = postMultipart($client, $apiUrl . '/pdf-with-added-tables', $apiKey, ['id' => $currentId, 'table_objects' => json_encode(tableObjects($metadata, $style, $items)), 'tag_enabled' => 'true', 'tag_language' => 'en-US'])['outputId'];
$currentId = postMultipart($client, $apiUrl . '/pdf-with-added-image', $apiKey, ['id' => $currentId, 'image_objects' => json_encode(['image_index' => 0, 'page' => 1, 'x' => 54, 'y' => 716, 'width' => 200, 'tag_alt_text' => 'Northstar Sample Supply logo', 'tag_structure_type' => 'Figure']), 'image_files' => ['path' => $dataDir . '/northstar-logo.png', 'name' => 'northstar-logo.png'], 'tag_enabled' => 'true', 'tag_language' => 'en-US'])['outputId'];
$pageCount = (int)postMultipart($client, $apiUrl . '/pdf-info', $apiKey, ['id' => $currentId, 'queries' => 'page_count'])['page_count'];
$currentId = postMultipart($client, $apiUrl . '/pdf-with-added-shapes', $apiKey, ['id' => $currentId, 'shape_objects' => json_encode(shapes($style, $pageCount, true)), 'tag_enabled' => 'true'])['outputId'];
$footer = [];
$addFooter = function (string $text, int $page, int $x, int $y, float $size, int $width, string $structure = 'P', bool $bold = false) use (&$footer, $style): void { $footer[] = ['font' => $bold ? $style['boldFont'] : $style['bodyFont'], 'max_width' => $width, 'opacity' => '1', 'page' => (string)$page, 'rotation' => '0', 'text' => $text, 'text_color_rgb' => rgb($style['mutedTextColorRgb']), 'text_size' => $size, 'x' => $x, 'y' => $y, 'tag_structure_type' => $structure]; };
$addFooter('Payment terms', $pageCount, 66, 156, 8, 480, 'H2', true); $addFooter($metadata['paymentTerms'], $pageCount, 66, 142, 7.5, 480); $addFooter('Notes', $pageCount, 66, 112, 8, 480, 'H2', true); $addFooter($metadata['notes'], $pageCount, 66, 98, 7.5, 480);
for ($page = 1; $page <= $pageCount; $page++) { $addFooter('Generated from structured JSON and CSV input with pdfRest.', $page, 54, 54, 7.5, 400); $addFooter("Page $page of $pageCount", $page, 490, 54, 7.5, 68); }
$final = postMultipart($client, $apiUrl . '/pdf-with-added-text', $apiKey, ['id' => $currentId, 'text_objects' => json_encode($footer), 'tag_enabled' => 'true', 'tag_language' => 'en-US', 'output' => 'invoice_from_structured_data']);
$pdf = $client->get($apiUrl . '/resource/' . rawurlencode($final['outputId']) . '?format=file', ['headers' => ['Api-Key' => $apiKey]]);
file_put_contents($outputPath, $pdf->getBody()->getContents());
echo "Created $outputPath\n";
